pubsys: Do not allow private AMIs to be published to SSM parameters in the public namespace

cbgbt opened the pull request:
**Description of changes:**
This change by default will disallow publishing SSM parameters to the public `/aws/` SSM namespace if they refer to private AMIs.



**Testing done:**

* Here's a sample run where I tried to set some of my own private AMIs to the `/aws/service/doggierocket` namespace. ( :dog: )
```
03:58:05 [INFO] Using AMI data from path: ../../build/images/aarch64-aws-k8s-1.21/bottlerocket-aws-k8s-1.21-aarch64-1.10.0-867fbbda-amis.json
03:58:05 [INFO] Parsing SSM parameter templates from policies/ssm/defaults.toml
03:58:05 [INFO] Ensuring that only public images are published to public parameters.
03:58:05 [ERROR] Attempted to set parameter '/aws/service/doggierocket/aws-k8s-1.21/arm64/1.10.0/image_version' in us-west-2 to '1.10.0', based on AMI ami-0bc2a2c72b2ce0a28. That AMI is not marked public!
03:58:05 [ERROR] Attempted to set parameter '/aws/service/doggierocket/aws-k8s-1.21/arm64/1.10.0/image_id' in us-west-2 to 'ami-0bc2a2c72b2ce0a28', based on AMI ami-0bc2a2c72b2ce0a28. That AMI is not marked public!
Failed to update SSM: Cowardly refusing to publish private image to public namespace without ALLOW_PRIVATE_IMAGES

[ec2-user@ip-10-0-0-27 pubsys]$ echo $?
1
```

* Using `--allow-private-images` passes this check (but then fails because we are trying to update parameters that don't exist)

Originally I sought to use Traits and fakes to provide additional automated testing here, but it was pretty difficult to do it in a way that didn't require more significant refactoring.



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

cbgbt committed:
pubsys-config: fix clippy warnings

cbgbt committed:
pubsys: fix clippy warnings

cbgbt committed:
pubsys: disallow private AMIs in public SSM params

stmcginnis commented on tools/pubsys/src/aws/ssm/mod.rs:
```diff
@@ -156,8 +210,56 @@ pub(crate) async fn run(args: &Args, ssm_args: &SsmArgs) -> Result<()> {
 }
 
+// Determines the number of AMIs that can be checked for public permissions in parallel.
+const CONCURRENT_AMI_CHECKS: usize = 16;
```
Would it be better to set this at runtime based on the running host's number of cores?

cbgbt replied:
Since this is network-bound, I picked an arbitrarily "low" number of concurrent checks. I've found that you often hit throttle limits trying to hit AWS using all of your cores if your host is too beefy. This can be rectified with more careful throttling avoidance, but I opted for a simpler route here.

I should probably clarify why the number was chosen in a comment.

stmcginnis replied:
Interesting, I was actually worried this number was too high. I was thinking something like `Min(cpu_cores, 16)`, but even wondering if that should be more like `8` than `16`.

cbgbt replied:
`8` is probably a safer bet!

stmcginnis reviewed (approved):
Looks good to me once the build failure is addressed. Great to have this!

cbgbt commented:
* Fix clippy warning on large size variation in Enum variants
* Lower concurrent AMI query to a safer value to avoid throttling

yeazelm reviewed (approved):
I'm still learning Rust for concurrency things, so I didn't review those specifics, but otherwise it looks good to me.

etungsten commented on tools/pubsys/src/aws/ami/public.rs:
```diff
@@ -0,0 +1,67 @@
+            region: region.to_string(),
+        })?;
+
+    let all_returned_images = described_image.images().context(error::NoSuchImageSnafu {
+        ami_id: ami_id.to_string(),
+        region: region.to_string(),
+    })?;
```
nit: `NoSuchImage` isn't accurate when `images()` returns `None`.

If `images` is `None`, it means the request has failed and not that there is no images in the list. If there were no images, the list would just be `Some(empty)`. It's kind of a weird paradigm with the AWS Rust SDK due to how Rust error handling works since if the request fails the error gets caught earlier with the `Result<_, SdkError>`. Theoretically this could never be `None` if the call returns `Ok`.

etungsten replied:
My suggestion here is to just `unwrap_or_default`

etungsten commented on tools/pubsys/src/aws/ssm/mod.rs:
```diff
@@ -112,21 +123,64 @@ pub(crate) async fn run(args: &Args, ssm_args: &SsmArgs) -> Result<()> {
+    // Generate AWS Clients to use for the updates.
+    let mut param_update_ops: Vec<SsmParamUpdateOp> = Vec::new();
+    let mut aws_sdk_configs: HashMap<Region, SdkConfig> = HashMap::new();
+    let mut ssm_clients = HashMap::new();
```
This was `    let mut ssm_clients = HashMap::with_capacity(amis.len());` before. Any harm in still pre-reserving memory for the number the AMIs we know we're gonna need to publish parameters for? Same with the other collections, do we know ahead of time the needed capacity?

etungsten commented on tools/pubsys/src/aws/ssm/mod.rs:
```diff
@@ -156,8 +210,57 @@ pub(crate) async fn run(args: &Args, ssm_args: &SsmArgs) -> Result<()> {
+    // Concurrently check our input parameter updates...
+    let results: Vec<Result<bool>> = futures::stream::iter(public_namespace_updates)
+        .then(|update| async move { Ok(check_ami_public(update)) })
+        .try_buffer_unordered(CONCURRENT_AMI_CHECKS)
```
Since there's no wait involved, wouldn't we still potentially hit some API call rate limit and get throttled or get errors? The rate at which we reach out to EC2 API doesn't change much even if we try 8 at a time (provided that the calls return promptly)?

cbgbt replied:
Good point. I looked into it, and rate-limiting a stream was actually pretty easy with the `governor` crate, so I went ahead and implemented it.

etungsten commented on tools/pubsys/src/aws/ssm/mod.rs:
```diff
@@ -356,4 +483,6 @@ mod error {
 }
 pub(crate) use error::Error;
+
+use self::template::RenderedParameter;
```
nit:  This should probably near the top with the other `use` declarations

etungsten commented on tools/pubsys/src/aws/ssm/template.rs:
```diff
@@ -70,13 +70,31 @@ pub(crate) fn get_parameters(
+/// A value which stores rendered SSM parameters alongside metadata used to render their templates
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub(crate) struct RenderedParameter {
+    pub(crate) ami: Image,
```
Can we just store the AMI-ID, or do we need everything else about the Image as well?

cbgbt replied:
We do currently only use the AMI ID, but I added the full AMI here so that the image could be used in the future while logging to provide context to any errors. Could be useful in cases where we want to understand why a template was rendered in a given way.

I'd be okay with removing it since we don't currently do that, but it seemed like a good idea to me.

etungsten reviewed (commented):
Nice! Just have a few non-blocking questions and small suggestions.

cbgbt commented:
Addressed @etungsten's feedback
* Cleaned up `Option` handling when calling EC2 DescribeImages
* Added proper rate limiting to DescribeImages calls via `governor` crate
* Pre-allocate memory for data structures with known sizes
* linting fixes

cbgbt commented:
* Re-added `BSD-2-Clause` to `deny.toml`. This license is allowed, it was just previously not used.
